snuba/cli/dlq_manager.py

```diff
@@ -0,0 +1,194 @@
+import os
+from typing import MutableSequence, Optional, Sequence, Tuple
+
+import click
+from arroyo import Message, Partition, Topic
+from arroyo.backends.kafka import KafkaConsumer, KafkaPayload
+
+from snuba.datasets.storages.factory import WRITABLE_STORAGES
+from snuba.utils.streams.configuration_builder import build_kafka_consumer_configuration
+
+__STORAGE_TO_DLQ_TOPIC = {
+    storage: storage.get_table_writer()
+    .get_stream_loader()
+    .get_dead_letter_queue_topic()
+    for storage in WRITABLE_STORAGES.values()
+}
+
+STORAGE_SETS_WITH_DLQ = {
+    storage.get_storage_set_key().value: topic
+    for storage, topic in __STORAGE_TO_DLQ_TOPIC.items()
+    if topic is not None
+}
+
+LINE_BREAK = "-" * 50
+
+
+@click.group()
+def dlq_manager() -> None:
+    """
+    Tools for Snuba's Dead Letter Queues
+    """
+    pass
+
+
+@dlq_manager.command()
+@click.option(
+    "--storage-set",
+    "storage_set",
+    type=click.Choice([storage_set for storage_set in STORAGE_SETS_WITH_DLQ]),
+    help="The storage set to list dead letters for",
+    required=True,
+)
+@click.option(
+    "--offset",
+    "offset",
+    type=int,
+    help="An offset on the dead-letter topic to start reading from, earliest by default",
+    default=0,
+)
+@click.option(
+    "--partition",
+    "partition_index",
+    type=int,
+    help="The partition to read from, must be used with the --offset arg. See 'info' command.",
+    default=0,
+)
+@click.option(
+    "--limit",
+    "limit",
+    type=int,
+    help="Max number of messages to display, 10 by default",
+    default=10,
+)
+def list(storage_set: str, offset: int, partition_index: int, limit: int) -> None:
+    """
+    List all messages found in a dead-letter topic
+    """
+    click.echo("\nthis could take up to 10 seconds...\n")
+    messages = _consume_dead_letters(storage_set, offset, partition_index, limit)
+    if messages:
+        click.echo(f"Listing the first {limit} messages ", nl=False)
+
+        if offset != 0:
+            click.echo(f"from offset {offset} ", nl=False)
+        if partition_index != 0:
+            click.echo(f"on partition {partition_index} ", nl=False)
+
+        click.echo(f"in {storage_set} dead-letter topic:")
+        click.echo("(see --help for pagination)\n")
+
+        for message in messages:
+            click.echo(LINE_BREAK)
+            click.echo(f"DLQ Offset: {message.offset}\n")
+            click.echo(message.payload.value)
+    else:
+        click.echo(f"\nNo messages found in {storage_set} dead-letter topic!")
+    click.echo(LINE_BREAK)
+
+
+def _consume_dead_letters(
+    storage_set: str, offset: int, partition_index: int, limit: int
+) -> Sequence[Message[KafkaPayload]]:
+    consumer = _build_consumer(storage_set)
+    message = consumer.poll(10)
```

Comment:
I think we need to clarify the multiple partition scenario, there is no guarantee this message is from the partition index being passed in

Reply:
Well that's what the info command would help with. It would list first+last available offset, and then a list of partitions with whatever offset they contain up to.

And if a bad partition index is passed in, the error message would be displayed.

Reply:
If you pass a partition and offset in though, you can get messages from another partition and offset entirely - is that expected? You can also get different results each time because there is no ordering between partitions.